<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>ASURI: An R Package for Survival Analysis and Risk Prediction using Gene Expression • asuri</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="ASURI: An R Package for Survival Analysis and Risk Prediction using Gene Expression">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">asuri</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/asuri.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jdelasrivas-lab/asuri/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>ASURI: An R Package for Survival Analysis and Risk Prediction using Gene Expression</h1>
            <h3 data-toc-skip class="subtitle">Survival analysis and
risk prediction for patients based on gene expression profiling combined
with disease survival time data.</h3>
                        <h4 data-toc-skip class="author">Alberto
Berral-Gonzalez</h4>
            <address class="author_afil">
      Bioinformatics and Functional Genomics Group - Cancer Research
Centre (CiC-IBMCC, CSIC/USAL) and IBSAL, Salamanca
(Spain)<br><h4 data-toc-skip class="author">Santiago
Bueno-Fortes</h4>
            <address class="author_afil">
      Bioinformatics and Functional Genomics Group - Cancer Research
Centre (CiC-IBMCC, CSIC/USAL) and IBSAL, Salamanca
(Spain)<br><h4 data-toc-skip class="author">Jose
Manuel Sanchez-Santos</h4>
            <address class="author_afil">
      Bioinformatics and Functional Genomics Group - Cancer Research
Centre (CiC-IBMCC, CSIC/USAL) and IBSAL, Salamanca
(Spain)<br><h4 data-toc-skip class="author">Manuel
Martin-Merino</h4>
            <address class="author_afil">
      Bioinformatics and Functional Genomics Group - Cancer Research
Centre (CiC-IBMCC, CSIC/USAL) and IBSAL, Salamanca
(Spain)<br><h4 data-toc-skip class="author">Javier
De Las Rivas</h4>
            <address class="author_afil">
      Bioinformatics and Functional Genomics Group - Cancer Research
Centre (CiC-IBMCC, CSIC/USAL) and IBSAL, Salamanca
(Spain)<br><small class="dont-index">Source: <a href="https://github.com/jdelasrivas-lab/asuri/blob/HEAD/vignettes/asuri.Rmd" class="external-link"><code>vignettes/asuri.Rmd</code></a></small>
      <div class="d-none name"><code>asuri.Rmd</code></div>
    </address>
</address>
</address>
</address>
</address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p><strong>Introduction</strong>: Modern genomic medicine, based
      on omic technologies, provides a novel approach to the study of
      diseases, since it can facilitate the identification of molecular
      biomarkers associated with risk, prognosis or outcome. To achieve
      this, we need accurate methods to analyze survival data from
      patient cohorts and correlate these data with the expression level
      of specific genes associated with the disease (i.e., in the case
      of cancer, associated with tumor prognosis). In addition, this
      type of analysis can also generate data for patient risk
      prediction and patient stratification, which are critical for a
      real progress in personalized medicine. The ASURI package
      presented here provides a set of integrated tools for survival
      analysis and patient risk prediction based on gene expression
      profiling, ensuring the discovery of novel gene survival markers
      with high robustness and reproducibility.</p>
      <p><strong>Key Features</strong>:</p>
      <ol style="list-style-type: decimal">
<li><p>Evaluate the association of gene expression with survival
      time (<strong>geneSurv</strong>).</p></li>
      <li><p>Discover gene markers that correlate with phenotypic
      factors (<strong>genePheno</strong>).</p></li>
      <li><p>Build robust patient risk predictors using genes associated
      with survival (<strong>patientRisk</strong>).</p></li>
      <li><p>Predict patient risk and classify into risk groups
      (<strong>predict.patientRisk</strong>).</p></li>
      <li><p>Estimate survival for novel query patients
      (<strong>predSurvCurve</strong>).</p></li>
      </ol>
</div>
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Here we present a new algorithm called <strong>ASURI</strong> that
provides an integrated set of functions to analyse disease
<strong>SURVIVAL</strong> and obtain patient <strong>RISK</strong>
predictions based on gene activity signatures. Specifically, the
algorithm is designed to: (1) Analyse the ability of a given gene to
mark disease prognosis and survival based on gene expression or other
gene-related activity signal; (2) Discover gene markers by identifying
the significant association of gene expression (or other gene-related
signal) with a clinical variable or phenotypic trait; (3) Constructing
robust patient risk predictors based on selected gene signatures using
univariate and multivariate approaches; (4) Predicting the risk of new
query patients based on their gene expression profile and classifying
the patients into a risk group (high/intermediate/low); (5) Estimating a
survival curve for new query patients.</p>
<p>In accordance with the 5 utilities of the package described above,
<strong>ASURI</strong> is built with five R functions specific to each
of these uses (<strong>Figure @ref(fig:diagramaCrop)</strong>): (1)
<strong>geneSurv()</strong>, that implements a robust method to evaluate
the ability of a given gene to mark survival based on its expression
level. It is based on a bootstrapped version of the Kaplan-Meier
estimator and the long-rank statistics provided by <span class="citation">(Therneau 2021)</span>. This function helps to overcome
the instability and irreproducibility of other tools developed for this
purpose <span class="citation">(Aguirre-Gamboa et al. 2013)</span>. (2)
<strong>genePheno()</strong>, which implements a robust method to select
a subset of genes whose expression is correlated with a clinical
variable or phenotypic parameter. This function is based on the Elastic
Net algorithm proposed by <span class="citation">(Friedman, Hastie, and
Tibshirani 2010; Simon et al. 2011)</span>, which removes features that
are not-relevant to the prediction the clinical variable. To improve the
stability and reproducibility of the list of genes associated with a
phenotype, the method implements a bootstrap strategy. The function
provides several metrics that are useful to evaluate the stability and
relevance of each gene. This function is particularly useful for
clinicians who want to discover survival biomarkers that interact with
clinical variables or parameters relevant to patient treatment.</p>
<div class="figure" style="text-align: center">
<img src="figures%2FdiagramaCrop.png" alt="Workflow diagram of the package ASURI showing the decision-making lines and the five main functions." width="100%"><p class="caption">
Workflow diagram of the package ASURI showing the decision-making lines
and the five main functions.
</p>
</div>
<p>Next function, (3) <strong>patientRisk()</strong>, which accurately
predicts the patient risk given a set of informative genes, provided as
an input list of selected genes. The function is based on the
multivariate UNICOX model proposed by <span class="citation">(R.
Tibshirani 2009)</span>. The function provides as output the list of
selected genes ranked according to their impact on patient risk along
with a relevance index. It also provides a risk prediction for each one
of the patients included in the training considering the gene signature
obtained and an optimal stratification into two or three risk groups
(i.e., poor, intermediate and good prognosis). (4)
<strong>predict.patientRisk()</strong>, which predicts the risk for new,
novel patients based on the model and the gene signature produced by the
<strong>patientRisk()</strong> function over the training set. This
function also classifies the patients in two or three risk groups. (5)
<strong>predSurvCurve()</strong>, which estimates the survival curve for
any given query patient using the patient’s gene expression profile.
These 5 functions included in ASURI also provide graphical output plots
to help visualize and analyze survival and patient risk; and, as
indicated, they are designed to work with one or more genes (i.e.,
features), that is, to analyze univariate and multivariate gene
signatures.</p>
</div>
<div class="section level2">
<h2 id="package-download-and-installation--package-dependencies-">Package download and installation. Package dependencies.<a class="anchor" aria-label="anchor" href="#package-download-and-installation--package-dependencies-"></a>
</h2>
<p>The ASURI package can be installed directly from Bioconductor typing
the following commands in an R:</p>
<p>if (!requireNamespace(“BiocManager”, quietly = TRUE)) {
install.packages(“BiocManager”) }</p>
<p>BiocManager::install(“ASURI”)</p>
<p>It can be installed also from a local directory if you have
downloaded previously the package.</p>
<p>install.packages(“/path/to/ASURI.tar.gz”, dependencies = TRUE)
library(ASURI)</p>
<p>If a problem with dependencies is reported during installation, users
should first download the missing R packages directly from <a href="https://cran.r-project.org/" class="external-link">CRAN</a> R repository or from <a href="https://bioconductor.org/" class="external-link">Bioconductor</a> repository, and then
repeat previous instructions to install ASURI R package. In Windows
based OS, <a href="https://cran.r-project.org/bin/windows/Rtools/" class="external-link">Rtools</a> might
need to be installed. Some of the package dependencies of ASURI are the
following, from <strong>CRAN</strong>: ‘ggplot2’, ‘glmnet’, ‘Rdpack’,
‘ROCR’, ‘scales’; and from <strong>Bioconductor</strong>: ‘Biobase’,
‘siggenes’, ‘survcomp’. The ASURI package also contains an experimental
dataset to illustrate the performance of the various functions and
methods implemented in it.</p>
</div>
<div class="section level2">
<h2 id="dataset-description-and-download-">Dataset description and download.<a class="anchor" aria-label="anchor" href="#dataset-description-and-download-"></a>
</h2>
<p>The dataset included in ASURI as a study case is intended to
facilitate the use of the package. The dataset was obtained from several
GSE series from <a href="http://www.ncbi.nlm.nih.gov/geo/" class="external-link">GEO
database</a> corresponding to breast cancer (BRCC) samples, where
expression and survival time were reported <span class="citation">(Bueno-Fortes et al. 2023)</span>. Each sample
corresponds to genome-wide expression profiles of BRCC primary tumor
samples hybridized on the transcriptomic platform: <em>Affymetrix</em>
<em>HGU133 Plus2.0</em>; which is a high-density microarray expression
platform containing 594,000 oligo-nucleotide probes (organized in
probe-sets) that we mapped to ENSEMBL genes using the
<em>hgu133plus2hsensgcdf</em> CDF package, obtained from <a href="http://mbni.org/customcdf/24.0.0/ensg.download/hgu133plus2hsensgcdf_24.0.0.tar.gz" class="external-link">BRAINARRAY</a>.
The expression signal of the samples was normalized using fRMA <span class="citation">(McCall, Bolstad, and Irizarry 2010)</span> or RMA
<span class="citation">(Gautier et al. 2004)</span> and Combat <span class="citation">(Kupfer et al. 2012)</span>, as described in <span class="citation">(Bueno-Fortes et al. 2023)</span>. Clinical variables
and information about the tumors were collected and included in the
phenodata object. More specifically, the information about the
expression level of 3 key BRCC biomarkers (i.e. <strong>ER</strong> /
ESR1, <strong>PR</strong> / PGR, and <strong>HER2</strong> / ERBB2
protein-coding genes) was included in the phenodata when available.</p>
<p>Once this dataset is downloaded two objects are created:
<em>mExprs</em>, that is the normalized gene expression data matrix; and
<em>mPheno</em> that includes information about each sample
(i.e. clinical physio-pathological data and other phenotypic data).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">mExprs</span><span class="op">)</span></span>
<span><span class="co"># To visualize the dimensions of the ExpressionSet object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">mExprs</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">mPheno</span><span class="op">)</span></span>
<span><span class="co"># To visualize the dimensions of the phenotypic information</span></span>
<span><span class="co"># object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">mPheno</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="platforms-supported-by-the-library-and-preprocessing-">Platforms supported by the library and preprocessing.<a class="anchor" aria-label="anchor" href="#platforms-supported-by-the-library-and-preprocessing-"></a>
</h2>
<p>The library supports expression data matrices coming from
high-density microarrays (such as the example data set described above),
as well as expression data matrices coming from RNA-seq platforms.
However, it can be applied to any omic technology that provides a
gene-related activity signal.</p>
<p>For the transcriptomic / gene expression data, it is recommended to
transform the gene signal using the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>l</mi><mi>o</mi><msub><mi>g</mi><mn>2</mn></msub></mrow><annotation encoding="application/x-tex">log_2</annotation></semantics></math>
function and to standardize the expression values, for example by
calculating a z-score. For microarray data, it is recommended to perform
a robust normalization of the raw signal before running <em>COX
regression methods</em>. In addition, for the microarray datasets, the
gene expression signals are calculated using the <strong>RMA</strong>
method <span class="citation">(Gautier et al. 2004; McCall, Bolstad, and
Irizarry 2010)</span> or the <strong>normalizeBetweenArrays</strong>
method from <span class="citation">(Ritchie et al. 2015)</span> (both
work well within the ASURI package). For RNA-seq datasets, the data
matrix of counts per gene can be used, and we recommend applying a
normalization method such as the Voom method, considered in <span class="citation">(Scarfo et al. 2016)</span>.</p>
</div>
<div class="section level2">
<h2 id="survival-and-risk-assessment-case-study-on-breast-cancer-">Survival and risk assessment (case study on Breast Cancer).<a class="anchor" aria-label="anchor" href="#survival-and-risk-assessment-case-study-on-breast-cancer-"></a>
</h2>
<p>In this section, we present and explain the use of the
<strong>ASURI</strong> library, including the five main functions
described above, by applying the package to the example breast cancer
dataset (which contains 200 samples and 20,049 genes). The different
parts of this main section show the different analyses that can be
performed with the package.</p>
<div class="section level3">
<h3 id="analysis-of-individual-genes-as-survival-markers-genesurv">Analysis of individual genes as survival markers:
<strong>geneSurv()</strong><a class="anchor" aria-label="anchor" href="#analysis-of-individual-genes-as-survival-markers-genesurv"></a>
</h3>
<p>The first goal of the <strong>ASURI</strong> algorithm is to identify
human genes measured in biological samples from patients (e.g., measured
in tumor samples from cancer patients) that are associated with the
survival of the individuals (i.e., with the prognosis). To do this, the
algorithm examines the expression (or activity) of the query genes in a
cohort of samples from patients with the same disease, and divides such
a cohort into samples with high expression of a query gene and samples
with low expression of such query gene. In this way, the algorithm
splits the cohort into two groups and then runs a Kaplan-Meier (K-M)
analysis for each of these two groups of samples, determining whether
such K-M show a significant difference based on the log-rank statistic.
All of this is implemented in the <strong>geneSurv()</strong> function
<span class="citation">(Bueno-Fortes et al. 2023)</span>. This statistic
evaluates the separability between the survival curves of the groups
obtained by a gene expression (or gene activity) threshold, which will
be: samples with high expression versus samples with low expression of
the query gene. The naive approach introduced in other tools, such as
<span class="citation">(Aguirre-Gamboa et al. 2013)</span>, estimates
the two groups of samples just splitting the cohort using the mean or
median of the gene activity, and then trying to obtain the best p-value
of the log-rank statistic for the KM of the two groups. This strategy
often yields meaningless solutions, giving rise to highly unbalanced
groups. To overcome these problems, the <strong>geneSurv()</strong>
function implements an original method that uses bootstrapping to
compute the optimal threshold and focuses the split only on the central
60% of the samples, avoiding very unbalanced groups.</p>
<p>As described above, to improve the stability and reproducibility of
the survival analysis available in other R packages <span class="citation">(Kassambara et al. 2024)</span>, a bootstrap strategy
is implemented in <strong>ASURI</strong>. A membership probability is
estimated for the classification of the each patient (each sample) into
one of the two risk groups (<strong>Figure
@ref(fig:pclassesr1)</strong>). These membership probabilities allow us
to robustly reclassify patients around the gene expression threshold
(<strong>Figure @ref(fig:boxesr1)</strong>). This approach is univariate
and does not take into account interactions or correlations between the
genes (i.e., it considers each feature gene as independent).</p>
<p>Once the patients are assigned to the two groups by the bootstrap
strategy, the function plots robust K-M curves with confidence intervals
(<strong>Figure @ref(fig:kmesr1)</strong>).</p>
<p>As an example of use, let’s now analyse using the
<strong>geneSurv()</strong> function a well known <strong>gene survival
marker</strong> in the dataset of breast cancer: the <strong>Estrogen
Receptor (ER+, ESR1)</strong>.</p>
<p>The input parameters for the function are the following:</p>
<ul>
<li>
<code>geneExpr</code> Vector with the normalized gene expression for
each sample. names(genExpr) should contains the sample names.</li>
<li>
<code>time</code> Numeric vector containing the survival time for
each sample in years, including the sample name in names(time).</li>
<li>
<code>status</code> Numeric vector with the status (censored 0 and
not censored 1) for each sample, names(status) should include the sample
names.</li>
<li>
<code>geneName</code> A character string with the name of the gene
being analyzed.</li>
<li>
<code>boxplot</code> A logical value indicating whether to generate
a boxplot of gene expression by survival group (by default = TRUE).</li>
<li>
<code>iter</code> The number of iterations (bootstrap resampling)
for calculating optimal group cutoffs (by default = 100).</li>
<li>
<code>type</code> Specifies whether the two Kaplan-Meier curves are
computed using the gene expression groups (by default “exprs”), or using
the risk groups (“risk”) if the samples are already classified by risk
level and this is used directly to divide the samples into two
groups.</li>
<li>
<code>cut_time</code> A numeric value specifying the cutoff time (in
years) for survival analysis. All events beyond this time are treated as
censored (by default = 10 years).</li>
</ul>
<p><img src="asuri_files/figure-html/unnamed-chunk-3-1.png" width="700"><img src="asuri_files/figure-html/unnamed-chunk-3-2.png" width="700"><img src="asuri_files/figure-html/unnamed-chunk-3-3.png" width="700"></p>
<p>The function returns different outputs, i.e., a list with values and
plots that is different depending on the specified <em>type</em> or
<em>mode</em> how it is run. Run of the function
<strong>geneSurv()</strong> as <code>type = "exprs"</code>:</p>
<ul>
<li>
<strong>geneName</strong>: A character string with the name of the
gene being analyzed.</li>
<li>
<strong>patientExpr</strong>: The expression level of the query gene
in each sample (i.e., in each patient).</li>
<li>
<strong>patientClass</strong>: Vector of group classification
according to the gene expression level: 2 = high expression (HighExpr)
and 1 = low expression level (LowExpr).</li>
<li>
<strong>patientClassProbality</strong>: Vector of membership
probabilities for the classification.</li>
<li>
<strong>wilcox.pvalue</strong>: The p-value from the Wilcoxon test
comparing the two expression groups.</li>
<li>
<strong>plot_values</strong>: A list containing the Kaplan-Meier
curve results, the log-rank test p-value, and the hazard ratio.</li>
</ul>
<p>Also three plots are provided:</p>
<ul>
<li>Plot of the probability (from 0 to 1) of the samples to belong to
one of the two groups or classes obtained based on the expression level
of the query gene (low expression and high expression) (<strong>Figure
@ref(fig:pclassesr1)</strong>).</li>
<li>Boxplots showing the expression distributions of the samples divided
into the two defined groups (LowExpr versus HighExpr). The plot includes
the p.value of the Wilcoxon test comparing these two expression groups
(<strong>Figure @ref(fig:boxesr1)</strong>).</li>
<li>Kaplan-Meier plot showing the curves calculated based on the
survival of the two groups of samples generated (i.e., samples with high
expression of the query gene, in red, and samples with low expression of
the query gene, in green). The plot includes the p.value of the log-rank
statistical test obtained when comparing the K-M survival profile of the
two groups (<strong>Figure @ref(fig:kmesr1)</strong>).</li>
</ul>
<p>For <code>type = "risk"</code>:</p>
<ul>
<li>
<strong>geneName</strong>: A character string with the name of the
gene being analyzed.</li>
<li>
<strong>patientExpr</strong>: The expression level of the query gene
in each sample (i.e., in each patient).</li>
<li>
<strong>risk_score_predicted</strong>: A numeric vector of predicted
or calculated risk score (in %, in relative numbers) of each
patient.</li>
<li>
<strong>plot_values</strong>: A list containing the Kaplan-Meier
curve results, the log-rank test p-value, and the hazard ratio.</li>
</ul>
<p>Also a plot is provided:</p>
<ul>
<li>Kaplan-Meier plot showing the curves calculated based on the
survival of the two groups of samples generated (high risk and low
risk). The plot includes the p.value of the log-rank statistical test
obtained when comparing the K-M survival profile of the two groups
(<strong>Figure @ref(fig:kmrisk)</strong>).</li>
</ul>
<div class="figure" style="text-align: center">
<img src="figures%2FPclassProbability.png" alt="Patient Class Probability for low exprs (89) and high exprs (111) groups)." width="80%"><p class="caption">
Patient Class Probability for low exprs (89) and high exprs (111)
groups).
</p>
</div>
<p><strong>Figure @ref(fig:pclassesr1)</strong> shows the plot of the
probability of the samples to be assigned to one of the expression
groups: low expression in green or high expression in red. The samples
in the example correspond to the tumors isolated from 200 breast cancer
patients. The probability of belonging to a group is estimated by
bootstrapping. Patients with a membership probability greater than 0.8
are classified with a high degree of confidence. The classification is
uncertain (p &lt; 0.8) for patients around the intermediate gene
expression level.</p>
<div class="figure" style="text-align: center">
<img src="figures%2Fexpr.png" alt="Boxplot: high expression vs low expression for ESR1 gene." width="80%"><p class="caption">
Boxplot: high expression vs low expression for ESR1 gene.
</p>
</div>
<p><strong>Figure @ref(fig:boxesr1)</strong> boxplots showing the
expression distribution of the samples divided into the two defined
groups (LowExpr versus HighExpr). The plot includes the p.value of the
Wilcoxon test comparing these two expression groups. Notice that there
is a small overlapping between the high and low expression groups. This
is due to the reassignment of patients with intermediate expression
level accomplished by the bootstrap strategy.</p>
<div class="figure" style="text-align: center">
<img src="figures%2FkmESR1.png" alt="Kaplan Meier curves obtained after the assignment of the patients by bootstrap to high or low survival using the selected feature ESR1." width="80%"><p class="caption">
Kaplan Meier curves obtained after the assignment of the patients by
bootstrap to high or low survival using the selected feature ESR1.
</p>
</div>
<p><strong>Figure @ref(fig:kmesr1)</strong> shows the two Kaplan-Meier
curves with the confidence intervals obtained by the function
<strong>geneSurv()</strong> for the gene ESR1. The threshold for the
gene expression level that splits the samples in two groups is estimated
by optimization of the log-rank test statistic. The procedure is
repeated 100 times using a bootstrap strategy to improve the
classification of intermediate risk patients.</p>
<p>The result shown in <strong>Figure @ref(fig:kmesr1)</strong>, with a
p-value for the log-rank test (= 0.01695) and an Hazard Ratio (= 1.86:
CI (0.31, 3.13) ) suggest that the gene ESR1 is a good survival marker
for breast cancer.</p>
<p>Run of the function <strong>geneSurv()</strong> as
<code>type = "risk"</code>:</p>
<p><img src="asuri_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<div class="figure" style="text-align: center">
<img src="figures%2Fkmtyperisk.png" alt="Kaplan Meier curves obtained after the assignment of the patients by bootstrap to high or low survival using the estimated risk based on feature BRCA1." width="80%"><p class="caption">
Kaplan Meier curves obtained after the assignment of the patients by
bootstrap to high or low survival using the estimated risk based on
feature BRCA1.
</p>
</div>
<p><strong>Figure @ref(fig:kmrisk)</strong> shows the two Kaplan-Meier
curves with the confidence intervals obtained by the function
<strong>geneSurv()</strong> for the gene BRCA1. Now the two groups are
obtained by estimating the risk for each patient and determining the
split threshold by optimizing the log-rank test statistic between the
two curves. The procedure is repeated 100 times using a bootstrap
strategy to improve the classification of intermediate risk
patients.</p>
</div>
<div class="section level3">
<h3 id="discovery-of-genes-associated-with-phenotypic-factors-genepheno">Discovery of genes associated with phenotypic factors:
<strong>genePheno()</strong><a class="anchor" aria-label="anchor" href="#discovery-of-genes-associated-with-phenotypic-factors-genepheno"></a>
</h3>
<p>It has been suggested in the literature that the clinical features
and physiopathological parameters of a patient’s disease can be
associated with specific biomolecular pathways or genes that can be
identified as markers of that patient’s condition or status. The
<strong>genePheno()</strong> function implements a robust algorithm to
obtain a list of genes associated with a given clinical variable or
phenotypic factor <span class="citation">(Bueno-Fortes et al.
2023)</span> measured in a cohort of patients. These genes may provide
alternative and novel targets to standard known disease markers.</p>
<p>The <strong>genePheno()</strong> function performs a feature
selection by fitting a predictor to the given set of samples based on
the <em>Elastic Net</em> algorithm (proposed by <span class="citation">(Friedman, Hastie, and Tibshirani 2010; Simon et al.
2011)</span>). This method incorporates a regularization term that lies
between the <em>Lasso</em> and the <em>Ridge</em> regression. Small
coefficients are reduced to zero but the predictor can include more
relevant variables than the Lasso to improve the prediction accuracy,
even though some of them may be slightly correlated. In this way, we
avoid the extreme behavior of Lasso, which randomly selects only one of
them. Thus, the <em>Elastic Net</em> algorithm helps to improve the
stability and to reduce the variance in the feature selection process
<span class="citation">(Robert Tibshirani 1996)</span>. The optimal
value for the regularization parameter is determined automatically by a
nested cross-validation strategy. The loss function for the predictor is
the AUC (Area Under the ROC Curve), which considers a balance between
false positive and false negative errors.</p>
<p>To improve the reproducibility and stability of the list of candidate
genes found (i.e., the selected features) found to be significantly
correlated with a clinical variable, a <strong>bootstrap</strong>
strategy is implemented. Training data are resampled with replacement
and an ensemble of gene lists is obtained. A stability metric based on
the selection probability is defined. Thus, genes with a low selection
probability are robustly filtered out. A final gene list is obtained
with two metrics. The first one evaluates the stability and the second
one the degree of correlation with the clinical variable. This approach
is multivariate and takes into account additive interactions between
candidate genes as well as possible correlations.</p>
<p>Before running <strong>genePheno()</strong>, the number of input
features, i.e. the number of tested genes, should be reduced (to no more
than a thousand) to improve the computational efficiency of the
function. However, this preprocessing step should be done carefully to
avoid removing relevant genes. The <strong>ASURI</strong> package
provides the function <strong>prefilterSAM()</strong>, which implements
a differential expression algorithm called SAM (proposed by <span class="citation">(Tusher, Tibshirani, and Chu 2001; Li and Tibshirani
2013)</span>) to obtain a set of significant genes to be selected.
<strong>prefilterSAM()</strong> improves the stability and robustness of
SAM by applying a bootstrapping strategy. For each preliminary gene
list, the optimal lambda threshold is calculated by repeating the
process with different FDR values and selecting the one that minimizes
the p-value. The final list contains only those genes that are present
in at least 10% of the iterations. This step can be skipped if you have
already selected a set of informative genes.</p>
<p>The input parameters for the function are the following:</p>
<ul>
<li>
<code>mExprs</code>: A data matrix where the rows are the genes and
the columns are the samples. Thus, each entry is the expression value of
a gene in a sample.</li>
<li>
<code>groups_vector</code>: A binary vector indicating the group to
which each sample is assigned.</li>
<li>
<code>FDRfilter</code>: A numeric value indicating the FDR threshold
for selecting significant genes. The default is FDR = 0.05.</li>
<li>
<code>iter</code>: The number of iterations for the bootstrapping.
Default is 100.</li>
<li>
<code>percentageFilter</code>: A numeric value indicating the
percentage of iterations a gene must appear in to be considered
significant. Default is 80.</li>
</ul>
<p>The output of the <code>DE_list_genes</code> function is a list with
the genes obtained from the differential expression analysis ordered by
the p-value (from lowest to highest p-value).</p>
<p>The following code illustrates the use of the
<strong>genePheno()</strong> function to discover genes that are
correlated with the <strong>Estrogen Receptor (ER+, ESR1)</strong>
measured by the clinicians (by the pathologists) as a phenotypic
variable of breast cancer tumors (i.e., to discover genes correlated
with the level of ESR1 protein detected by immunohistochemistry, IHC, in
the BRCC tumors).</p>
<p>The input parameters are the following:</p>
<ul>
<li>
<code>mExpr</code>: A matrix with normalized gene expression data.
Rows correspond to samples, and columns correspond to genes.
<code>rownames(mExpr)</code> should be set to sample names, and
<code>colnames(mExpr)</code> should be set to gene names. Since we
selected a subset of differentially expressed genes, the
<strong>genePheno()</strong> function is only run with these genes.
Therefore, from <code>mExpr</code> we generate a new expression
matrix:<code>tmExprsDE</code>, that only includes the genes that are
significant after the DE analysis.</li>
<li>
<code>vectorGroups</code>: Clinical or phenotypic variable tested.
It must be provided as a binary vector. In this particular study
<code>vectorGroups</code> is <code>mPheno$ER.IHC</code>.</li>
<li>
<code>vectorSampleID</code>: Vector containing the sample names
ordered as in the expression matrix.</li>
<li>
<code>iter</code>: Number of bootstrap iterations (default: 100).
This can be changed if the function takes too long to run.</li>
<li>
<code>numberOfFolds</code>: Number of folds to implement nested
cross-validation (default: 10).</li>
</ul>
<div id="tabla-1">

</div>
<table class="table">
<thead><tr class="header">
<th>Gene</th>
<th>stability</th>
<th>betasMedian</th>
<th>betasMean</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>NAT1</td>
<td>0.90</td>
<td>0.13152533</td>
<td>0.13359943</td>
</tr>
<tr class="even">
<td>ESR1</td>
<td>0.84</td>
<td>0.16505283</td>
<td>0.16619666</td>
</tr>
<tr class="odd">
<td>USP6NL</td>
<td>0.73</td>
<td>-0.30479925</td>
<td>-0.35228853</td>
</tr>
<tr class="even">
<td>SUSD3</td>
<td>0.68</td>
<td>0.07534488</td>
<td>0.08846058</td>
</tr>
<tr class="odd">
<td>PREX1</td>
<td>0.67</td>
<td>0.10376235</td>
<td>0.13750165</td>
</tr>
<tr class="even">
<td>TPBG</td>
<td>0.60</td>
<td>0.22454524</td>
<td>0.25966395</td>
</tr>
<tr class="odd">
<td>CA12</td>
<td>0.56</td>
<td>0.08063623</td>
<td>0.08739032</td>
</tr>
<tr class="even">
<td>AGR3</td>
<td>0.55</td>
<td>0.04043867</td>
<td>0.04232274</td>
</tr>
<tr class="odd">
<td>PROX1</td>
<td>0.54</td>
<td>-0.38767094</td>
<td>-0.43736144</td>
</tr>
<tr class="even">
<td>CALML5</td>
<td>0.53</td>
<td>-0.03846191</td>
<td>-0.04940331</td>
</tr>
</tbody>
</table>
<div align="center">
<p>
<b>Table 1</b>: List of genes by Stability.
</p>
</div>
<p>The output of the function is an object with the following
elements:</p>
<ul>
<li>
<strong>genes</strong>: A list of genes ranked according to the
degree of association with the clinical or phenotypic variable
tested.</li>
<li>
<strong>listCoeff</strong>: A list with the beta regression
coefficients and the AUC score for each bootstrap iteration.</li>
<li>
<strong>stability</strong>: Gene selection probability (0 to 1)
estimated by bootstrap (the number of times discovered over “n”
iterations).</li>
<li>
<strong>betasMedian</strong>: Median of the beta coefficients over
the B replicates.</li>
<li>
<strong>betasMean</strong>: Mean of the beta coefficients over the B
replicates.</li>
<li>
<strong>betasTable</strong>: Output table with genes ordered by
decreasing value of the stability, that includes the
<strong>genes</strong>, the <strong>stability</strong>, the
<strong>betasMedian</strong> and the <strong>betasMean</strong>.</li>
</ul>
<p>The results obtained in <a href="#tabla-1"><b>Table 1</b></a> show
that the expression of ESR1 gene is strongly associated with the
clinical marker ER (i.e., with the detection by IHC of Estrogen Receptor
active in the tumors). This will be expected, but it is a result that
supports the value of the developed method presented here. In addition,
the AGR3, SUSD3 or CA12 genes have recently been found in the breast
cancer literature in relation to the ER marker <span class="citation">(Barnett et al. 2008; Yu et al. 2015; Garczyk et al.
2015)</span>; and they are also found by our method with a good ranking
in the association with ER (i.e., they present high stability). These
genes are positively correlated, suggesting that they may play a role or
function in a similar way to the clinical marker ER. Therefore, these
genes can be proposed as alternative molecular targets in breast cancer,
becoming excellent candidates to investigate their value in diagnosis,
prognosis, and sensitivity to endocrine therapies.</p>
</div>
<div class="section level3">
<h3 id="identification-of-gene-survival-markers-patient-risk-prediction-patientrisk">Identification of gene survival markers &amp; patient risk
prediction: <strong>patientRisk()</strong><a class="anchor" aria-label="anchor" href="#identification-of-gene-survival-markers-patient-risk-prediction-patientrisk"></a>
</h3>
<p>When a preliminary set of candidate genes is proposed, the
<strong>ASURI</strong> algorithm is designed to explore these genes and
identify which ones are correlated with survival and risk, and therefore
are adequate prognostic markers (i.e., good gene survival markers). In
the case of the study presented above, using the
<strong>genePheno()</strong> function, the algorithm explores the data
to identify gene markers that correlate with the presence of ER+ as a
clinical variable. However, in a more general use with the
<strong>patientRisk()</strong> function, the algorithm evaluates a whole
set of genes with respect to the survival profiles and builds a
multivariate risk predictor, identifying which genes are able to
accurately predict patient risk.</p>
<p>The analysis to find a significant correlation between the expression
level of the genes and the survival of the patients is performed by the
function <strong>patientRisk()</strong> <span class="citation">(Bueno-Fortes et al. 2023)</span>. This function is
based on the <em>Univariate Shrinkage COX Model</em> (UNICOX) proposed
by <span class="citation">(R. Tibshirani 2009)</span>. This method
shrinks irrelevant coefficients to zero, taking into account a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>L</mi><mn>1</mn></msub><annotation encoding="application/x-tex">L_1</annotation></semantics></math>
norm penalty term. Features are penalized individually, allowing all the
genes correlated with risk to be retained at a given numerical strength.
This behavior differs from Lasso models <span class="citation">(Robert
Tibshirani 1996)</span>, where only one representative gene is randomly
chosen from the group of correlated genes. The
<strong>patientRisk()</strong> function estimates the regularization
parameter using a double nested cross-validation to provide robustness
in the feature selection. This cross-validation includes an inner
(first) loop to estimate the optimal regularization parameter and an
outer (second) loop to estimate the risk.</p>
<p>Once the risk curve is obtained, <strong>patientRisk()</strong>
implements an original method to stratify the patients into two main
risk groups. This method is similar to the one used in (<strong>Figure
@ref(fig:pclassesr1)</strong>), since it orders the patients by risk and
computes the p-values of the log-rank test for each possible split of
the patients in two groups with their corresponding survival. Then, it
calculates the interval corresponding to the lower 10th percentile of
the p-values. The extreme values for this interval provide the lower and
upper thresholds that stratify the patients into low, intermediate and
high risk groups (see <strong>Figure @ref(fig:plot1risk)</strong>).</p>
<p>The <strong>optimal threshold for dividing the patients into two
groups (low/high)</strong> can be obtained by three methods:
<strong>(1)</strong> The first approach estimates the threshold by
minimizing the p-value of the log-rank statistic, i.e. by maximizing the
separability between the K-M curves for the high and low risk groups.
When multiple local minima occur, this may be sample dependent and
unstable. <strong>(2)</strong> To improve the robustness a second method
is implemented that computes the threshold as the median of the lower
10th percentile log-rank p-values. The lower 10 percentile selects the
smallest values from the p-value distribution that correspond to
intermediate risk patients who are on the border between the two groups.
This interval is more robust than a single minimum and provides good
experimental results for a wide variety of problems tested. The median
threshold in this interval may change from one iteration to another
because the distribution of p-values for patients with intermediate risk
may change due to sample variation. Therefore, <strong>(3)</strong> a
bootstrapping strategy is implemented for these intermediate risk
patients, which helps to improve the robustness of the threshold
obtained and provides a probability of membership for each risk
group.</p>
<p>The following code illustrates how to apply the
<strong>patientRisk()</strong> function to discover/identify significant
gene survival markers and to accurately predict patient risk.</p>
<p><img src="asuri_files/figure-html/unnamed-chunk-8-1.png" width="700"><img src="asuri_files/figure-html/unnamed-chunk-8-2.png" width="700"><img src="asuri_files/figure-html/unnamed-chunk-8-3.png" width="700"><img src="asuri_files/figure-html/unnamed-chunk-8-4.png" width="700"><img src="asuri_files/figure-html/unnamed-chunk-8-5.png" width="700"></p>
<p>The input parameters of the <strong>patientRisk()</strong> function
are the following:</p>
<ul>
<li>
<code>mExpr</code> A numeric matrix of gene expression data where
each row represents a gene and each column represents a sample.</li>
<li>
<code>time</code> A numeric vector representing the survival time
for each sample, names(time) should be defined as the sample names.</li>
<li>
<code>status</code> A numeric vector representing the event status
(1 = event, 0 = censored) for each sample, names(status) should be
defined as the sample names.</li>
<li>
<code>group.vector</code> A numeric vector specifying predefined
risk groups for the patients. This is optional.</li>
<li>
<code>method</code> A string that determines the method considered
to stratified the patients in two risk groups according to the risk
score for the lower 10th percentile of the p-values: (1)
<code>min.pval</code> Defines risk groups based on the minimum p-value.
(2) <code>med.pval</code> Defines risk groups based on the median
p-value. This approach can be more robust when several local minima are
found. Finally, (3) <code>class.probs</code> Defines risk groups based
on the classification probabilities from the model. It provides also a
membership probability to each risk group. <code>class.probs</code> is
selected by default.</li>
<li>
<code>nboot</code> An integer specifying the number of bootstrap
iterations for risk score calculation. Default is 50.</li>
</ul>
<p>The <strong>patientRisk()</strong> function returns a list containing
the following objects:</p>
<ul>
<li>
<strong>cv_risk_score</strong>: Risk score prediction for the
training set using a double nested crossvalidated strategy.</li>
<li>
<strong>cv_normalized_risk</strong>: Normalized risk score in the
interval (0, 100).</li>
<li>
<strong>table_genes_selected</strong>: Data frame with the following
columns: (i) the names for the genes selected by the COX regression;
(ii) the beta coefficients for the optimal multivariate COX regression
fitted to the training set; (iii) the Hazard Ratio for each gene; and
(iv) the p-value for the univariate log-rank statistical test. Genes are
shown by descending order of the HR index.</li>
<li>
<strong>table_genes_selected_extended</strong>: Table with the same
format as table_genes_selected. A search for local minima within a 5%
range of the selected minimum is performed. The goal is expanding the
list of significant genes to improve biological interpretability, since
the Lasso penalty drastically reduces the number of significant
genes.</li>
<li>
<strong>model.optimalLambda</strong>: Object that contains the
parameters for the optimal multivariate COX regression.</li>
<li>
<strong>groups</strong>: Vector of classification of patients in two
risk groups: HighRisk (2) or LowRisk (1).</li>
<li>
<strong>riskThresholds</strong>: Thresholds that allows to stratify
the test patients in three groups according to the predicted risk score:
low, intermediate and high risk.</li>
<li>
<strong>range.risk</strong>: Range of the unscaled risk score in the
training set.</li>
<li>
<strong>listmodels</strong>: List of models tested for different
values of the regularization parameter
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>λ</mi><annotation encoding="application/x-tex">\lambda</annotation></semantics></math>.</li>
<li>
<strong>evaluation.models</strong>: Dataframe that provides several
metrics for each model evaluated. <strong>lambda</strong> is the
regularization parameter for the multivariate COX regression adjusted;
<strong>number_features</strong> gives the number of genes selected by
this model; <strong>c.index</strong> and <strong>se.c.index</strong> are
the concordance index and the standard deviation for the risk
prediction; and <strong>p_value_c.index</strong> and
<strong>logrank_p_value</strong> give the p-values for the the
concordance index and the log-rank statistics respectively. The models
are shown by ascending order of p-values of the log-rank test, and the
best one is marked with two asterisks. All the metrics have been
estimated using a double nested cross-validation strategy.</li>
<li>
<strong>betasplot</strong>: Dataset used to create the plot of genes
ranked according to the regression coefficients in the multivariate COX
model (UNICOX).</li>
<li>
<strong>plot_values</strong>: A list containing Kaplan-Meier fit
results, logrank p-value, and hazard ratio.</li>
<li>
<strong>membership_prob</strong>: If method <code>class.probs</code>
is selected a table with two columns is returned. The first one is the
probability of classification to the low risk group while the second one
is the membership probability to the high risk group.</li>
</ul>
<p>The function <strong>patientRisk()</strong> generates 5 plots:</p>
<div class="figure" style="text-align: center">
<img src="figures%2Fpval.png" alt="Curve of log-rank p-values for each possible splitting of patients into 2 risk groups. The X axis represents the patient number ranked from low to high risk. The green line marks the selected cutpoint, i.e., the point that provides the threshold for stratifying patients into 2 risk groups: low risk, in blue; high risk, in red. The blue &amp; red lines define thresholds for classifying patients into 3 risk groups: *low*, *intermediate* and *high*. These lines mark samples that fall in an *intermediate* region, where risk assignment is not very precise and samples may fluctuate in position." width="80%"><p class="caption">
Curve of log-rank p-values for each possible splitting of patients into
2 risk groups. The X axis represents the patient number ranked from low
to high risk. The green line marks the selected cutpoint, i.e., the
point that provides the threshold for stratifying patients into 2 risk
groups: low risk, in blue; high risk, in red. The blue &amp; red lines
define thresholds for classifying patients into 3 risk groups:
<em>low</em>, <em>intermediate</em> and <em>high</em>. These lines mark
samples that fall in an <em>intermediate</em> region, where risk
assignment is not very precise and samples may fluctuate in position.
</p>
</div>
<div class="figure" style="text-align: center">
<img src="figures%2Frisk.png" alt="Risk Score, between 0 and 100, calculated for each patient, i.e., for each of the 200 patient tumor samples included in this study, obtained with a multivariate UNICOX regression model. The model includes as features the subset of selected genes included in the *geneList*, which are tested to find out their correlation with risk. The green and blue &amp; red lines provide the thresholds estimated by the **patientRisk()** function to split the patients respectively in 2 or 3 groups, as described in the legend of the previous figure. Note that the implemented algorithm is capable of detecting changes in the slope of the risk curve, giving rise to risk groups of differenciated prognosis." width="80%"><p class="caption">
Risk Score, between 0 and 100, calculated for each patient, i.e., for
each of the 200 patient tumor samples included in this study, obtained
with a multivariate UNICOX regression model. The model includes as
features the subset of selected genes included in the <em>geneList</em>,
which are tested to find out their correlation with risk. The green and
blue &amp; red lines provide the thresholds estimated by the
<strong>patientRisk()</strong> function to split the patients
respectively in 2 or 3 groups, as described in the legend of the
previous figure. Note that the implemented algorithm is capable of
detecting changes in the slope of the risk curve, giving rise to risk
groups of differenciated prognosis.
</p>
</div>
<div class="figure" style="text-align: center">
<img src="figures%2Fpvaldistribution.png" alt="Log-rank p-values of the minimum optimal versus the number of genes selected by the model. As the regularization parameter in UNICOX increases, the regression model becomes more especific in the selection of features and, therefore the number of genes that have non-zero (non-null) coefficients is reduced." width="80%"><p class="caption">
Log-rank p-values of the minimum optimal versus the number of genes
selected by the model. As the regularization parameter in UNICOX
increases, the regression model becomes more especific in the selection
of features and, therefore the number of genes that have non-zero
(non-null) coefficients is reduced.
</p>
</div>
<p>The <strong>patientRisk()</strong> function offers multiple values
for the regularization parameter, the user can choose the value that
minimizes the log-rank p-value, and then re-evaluate everything with
that new gene list. For our case study the optimal value is achieved
with 12 genes. Biotechnologists, physicians, and researchers in general
often prefer to have a smaller set of selected genes (selected features)
as risk markers, at the expense of a slight increase in p-value. Taken
together, this representation allows us to choose the optimal number of
genes that achieve a balance between the lowest (most significant)
log-rank p-value and the fewest number of selected genes. Once the
subset of genes is selected, the corresponding COX model can be
retrieved from the list of objects returned by the
<strong>patientRisk()</strong> function. Finally, the risk score for new
patients (placed as an independent query) can be estimated considering
this constructed model as a predictor (this is further explained in the
next section). Note that the <strong>patientRisk()</strong> function can
be applied to any subset of genes provided by the user (related or
unrelated to a clinical variable). Thus, this function can be applied to
study the association of each gene with risk and to select the best gene
risk markers (i.e., providing a gene signature of risk markers). In this
study, considering the selected gene signature, the
<strong>patientRisk()</strong> function provides an accurate risk
prediction for each patient tumor sample and also the stratification of
patients in low, intermediate and high risk.</p>
<p><strong>Figure @ref(fig:plot4risk)</strong> presents the genes
selected by the multivariate COX model, ranked by the value of the
regression coefficients in the model. Genes with “positive” values (in
red) increase the hazard ratio (HR) and, therefore, increase the risk in
the samples (so they can be associated with an oncogenic effect). While
genes with “negative” values (in blue) reduce the HR and the risk in the
samples (so they can be associated with a tumor suppression or a
protective effect). The absolute value of the coefficients determines
the relevance of each gene with respect to the patient’s risk.</p>
<div class="figure" style="text-align: center">
<img src="figures%2Fbetas.png" alt="Plot of the genes selected by the multivariate COX model, ranked by the value of the regression coefficients in the model. Genes with *positive* values (red) increase the risk in the samples. While genes with *negative* values (blue) reduce the risk in the samples. Genes with significant p-values are marked with asterisks (** p-value&lt;0.01, * p-value&lt;0.05). Similarly, genes with slightly significance level are marked with a dot (· between 0.1 and 0.05). All the analyses carried out in this part are multivariate and take into account additive interactions among the genes." width="80%"><p class="caption">
Plot of the genes selected by the multivariate COX model, ranked by the
value of the regression coefficients in the model. Genes with
<em>positive</em> values (red) increase the risk in the samples. While
genes with <em>negative</em> values (blue) reduce the risk in the
samples. Genes with significant p-values are marked with asterisks (**
p-value&lt;0.01, * p-value&lt;0.05). Similarly, genes with slightly
significance level are marked with a dot (· between 0.1 and 0.05). All
the analyses carried out in this part are multivariate and take into
account additive interactions among the genes.
</p>
</div>
<table class="table">
<thead><tr class="header">
<th>Gene</th>
<th>Beta Value</th>
<th>Beta Group</th>
<th>P-Value</th>
<th>Signif</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>AKR1B10</td>
<td>0.32736318</td>
<td>type1</td>
<td>0.004322987</td>
<td>**</td>
</tr>
<tr class="even">
<td>EVL</td>
<td>-0.32328558</td>
<td>type2</td>
<td>0.010718153</td>
<td>*</td>
</tr>
<tr class="odd">
<td>SUSD3</td>
<td>-0.26224518</td>
<td>type2</td>
<td>0.003719841</td>
<td>**</td>
</tr>
<tr class="even">
<td>SLC39A6</td>
<td>-0.21327395</td>
<td>type2</td>
<td>0.016501657</td>
<td>*</td>
</tr>
<tr class="odd">
<td>ADIRF</td>
<td>-0.20193189</td>
<td>type2</td>
<td>0.022008641</td>
<td>*</td>
</tr>
<tr class="even">
<td>GFRA1</td>
<td>-0.14895182</td>
<td>type2</td>
<td>0.036662667</td>
<td>*</td>
</tr>
<tr class="odd">
<td>SOX11</td>
<td>0.14360541</td>
<td>type1</td>
<td>0.053214622</td>
<td>·</td>
</tr>
<tr class="even">
<td>CLIC6</td>
<td>-0.12819465</td>
<td>type2</td>
<td>0.057410691</td>
<td>·</td>
</tr>
<tr class="odd">
<td>CPB1</td>
<td>-0.11722549</td>
<td>type2</td>
<td>0.002609143</td>
<td>**</td>
</tr>
<tr class="even">
<td>PGR</td>
<td>-0.10142232</td>
<td>type2</td>
<td>0.199169763</td>
<td></td>
</tr>
<tr class="odd">
<td>NPY1R</td>
<td>-0.09605778</td>
<td>type2</td>
<td>0.027526344</td>
<td>*</td>
</tr>
<tr class="even">
<td>CALML5</td>
<td>0.07877817</td>
<td>type1</td>
<td>0.147320701</td>
<td></td>
</tr>
<tr class="odd">
<td>SERPINA5</td>
<td>-0.07173386</td>
<td>type2</td>
<td>0.280007397</td>
<td></td>
</tr>
<tr class="even">
<td>NAT1</td>
<td>-0.06950291</td>
<td>type2</td>
<td>0.253075375</td>
<td></td>
</tr>
<tr class="odd">
<td>SCGB1D2</td>
<td>-0.06748915</td>
<td>type2</td>
<td>0.114389857</td>
<td></td>
</tr>
<tr class="even">
<td>AGR3</td>
<td>-0.02526576</td>
<td>type2</td>
<td>0.548559368</td>
<td></td>
</tr>
</tbody>
</table>
<div align="center">
<p>
<b>Table 2</b>: List of genes by relevance for risk assignement.
</p>
</div>
<div class="figure" style="text-align: center">
<img src="figures%2Fkm.png" alt="Kaplan Meier plot: Blue (low risk) and red (high risk) of the two groups of samples with different survival obtained from the stratification of the multivariate risk curve (Figure 6)." width="80%"><p class="caption">
Kaplan Meier plot: Blue (low risk) and red (high risk) of the two groups
of samples with different survival obtained from the stratification of
the multivariate risk curve (Figure 6).
</p>
</div>
</div>
<div class="section level3">
<h3 id="risk-prediction-for-an-independent-set-of-patients-predict-patientrisk">Risk Prediction for an independent set of patients:
<strong>predict.patientRisk()</strong><a class="anchor" aria-label="anchor" href="#risk-prediction-for-an-independent-set-of-patients-predict-patientrisk"></a>
</h3>
<p>A relevant goal for medical doctors is to predict the risk of cancer
patients considering the level of certain molecular markers. This
prediction is implemented by function
<strong>predict.patientRisk()</strong> using a multivariate model that
takes into account the cooperative work between the genes to carry out a
biological function. Once the multivariate risk predictor is trained by
the <strong>patientRisk()</strong> function (see previous section), the
predict.patientRisk function may be applied to predict the risk of new
patients considering the gene expression of a small subset of genes
associated to risk. Moreover, the patients can be classified into low,
intermediate or high risk groups considering the value of the risk
score.</p>
<p>The following lines show how to predict the risk score for an
independent set of patients. The function plots the risk score and
stratifies the patients into three groups.</p>
<p><img src="asuri_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>The input parameters of the function
<strong>predict.patientRisk()</strong> are:</p>
<ul>
<li>
<code>model.fit</code> A list containing the pre-fitted model and
necessary parameters for risk prediction, including the optimal lambda
value, risk thresholds, and plot values given by
<strong>patientRisk()</strong> function.</li>
<li>
<code>mExpr.testData</code> A data frame for an independent set of
patients representing the gene expression data, where each row is a gene
and each column is a sample. The genes must be the same subset of genes
(as rows) considered in the training of the COX model using patientRisk
function. <em>colnames(mExpr)</em> corresponds to the sample names and
<em>rownames(mExpr)</em> to the gene names.</li>
</ul>
<p>The output of the function <strong>predict.patientRisk()</strong> is
a risk plot and a list with two objects:</p>
<ul>
<li>
<strong>risk_score</strong>: The unscaled risk score provided by the
multivariate COX regression.</li>
<li>
<strong>scaled_risk_score</strong>: The scaled risk score (0-100)
for the independent set of patients.</li>
</ul>
This function provides a plot with the predicted risk score using the
multivariate COX model trained by function
<strong>patientRisk()</strong>.
<div class="figure" style="text-align: center">
<img src="figures%2Frisk_plot_predict_patientRisk.png" alt="Risk score prediction for an independent set of patients using the optimal multivariate COX model trained by function patientRisk(). Green and blue/red lines correspond to the optimal thresholds that split the patients in two and three groups respectively. These thresholds were learned in training phase." width="80%"><p class="caption">
Risk score prediction for an independent set of patients using the
optimal multivariate COX model trained by function patientRisk(). Green
and blue/red lines correspond to the optimal thresholds that split the
patients in two and three groups respectively. These thresholds were
learned in training phase.
</p>
</div>
<p>This function can be applied also to predict the risk of a single
patient given the gene expression profile. The mExprs_testData will
contain now a column with the expression level for the list of genes
selected.</p>
<p>When a single patient is considered, the function will print the
normalized risk for the test patient and the classification into three
possible groups: Low, Intermediate or High. Treatment for the
intermediate group is uncertain and should be studied with care.</p>
</div>
<div class="section level3">
<h3 id="survival-estimation-from-the-cox-regression-for-a-test-patient-predsurvcurve">Survival estimation from the COX regression for a test patient:
<strong>predSurvCurve()</strong><a class="anchor" aria-label="anchor" href="#survival-estimation-from-the-cox-regression-for-a-test-patient-predsurvcurve"></a>
</h3>
<p>The function <strong>predSurvCurve()</strong> will obtain the
expected survival curve for a test patient estimated from the risk score
predicted by a multivariate COX regression model and considering the
Breslow estimator for the baseline Hazard. The function provides a plot
for the expected survival curve.</p>
<p>The input parameters are:</p>
<ul>
<li>
<code>cox_pred_training</code> A numeric vector representing the
predicted risk scores (or log-risk scores) from the COX model for the
training set. It can be obtained from
<strong>predict.patientRisk()</strong> function.</li>
<li>
<code>mSurv</code> A data frame with two columns: “time”
representing survival times, and “status” representing the event status
(1 for event, 0 for censored).</li>
<li>
<code>cox_pred_test</code> A numeric vector of length 1 representing
the predicted risk score (or log-risk score) for the test patient. It
can be obtained from <strong>predict.patientRisk()</strong>
function.</li>
<li>
<code>eval_surv_times</code> A numeric vector of times at which the
survival curve should be evaluated. If NULL (default), the times will be
taken from the training data up to the maximum event time.</li>
</ul>
<p><img src="asuri_files/figure-html/unnamed-chunk-12-1.png" width="700"><img src="asuri_files/figure-html/unnamed-chunk-12-2.png" width="700"></p>
<p>The output for this function is a list with three objects:</p>
<ul>
<li>
<strong>eval_times</strong>: Vector with the times at which the
survival curve has been evaluated.</li>
<li>
<strong>S_0_t</strong>: Vector with the baseline survival
probability estimated at each <code>eval_times</code>.</li>
<li>
<strong>S_test_patient</strong>: Vector with the survival
probability at <code>eval_times</code> for the test patient.</li>
</ul>
It provides also a plot with the expected survival curve estimation for
the test patient.
<div class="figure" style="text-align: center">
<img src="figures%2Fsurv_plot_breslow.png" alt="Survival curve estimation for a single patient based on the risk score predicted by a multivariate COX model. The blue line shows the median survival time at which the survival probability drops to 0.5." width="80%"><p class="caption">
Survival curve estimation for a single patient based on the risk score
predicted by a multivariate COX model. The blue line shows the median
survival time at which the survival probability drops to 0.5.
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">## <span style="color: #00BBBB; font-weight: bold;">─ Session info ───────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">##  <span style="color: #555555; font-style: italic;">setting </span> <span style="color: #555555; font-style: italic;">value</span></span></span>
<span><span class="co">##  version  R version 4.4.3 (2025-02-28)</span></span>
<span><span class="co">##  os       Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">##  system   x86_64, linux-gnu</span></span>
<span><span class="co">##  ui       X11</span></span>
<span><span class="co">##  language en</span></span>
<span><span class="co">##  collate  es_ES.UTF-8</span></span>
<span><span class="co">##  ctype    es_ES.UTF-8</span></span>
<span><span class="co">##  tz       Europe/Madrid</span></span>
<span><span class="co">##  date     2025-04-24</span></span>
<span><span class="co">##  pandoc   3.1.3 @ /usr/bin/ (via rmarkdown)</span></span>
<span><span class="co">##  quarto   1.3.353 @ /usr/lib/rstudio/resources/app/bin/quarto/bin/quarto</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB; font-weight: bold;">─ Packages ───────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">##  <span style="color: #555555; font-style: italic;">package     </span> <span style="color: #555555; font-style: italic;">*</span> <span style="color: #555555; font-style: italic;">version   </span> <span style="color: #555555; font-style: italic;">date (UTC)</span> <span style="color: #555555; font-style: italic;">lib</span> <span style="color: #555555; font-style: italic;">source</span></span></span>
<span><span class="co">##  asuri        * 0.99.0     <span style="color: #555555;">2025-04-24</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  Biobase      * 2.66.0     <span style="color: #555555;">2025-03-07</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">bioc_git2r (@2cd604f)</span></span></span>
<span><span class="co">##  BiocGenerics * 0.52.0     <span style="color: #555555;">2025-03-07</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">bioc_git2r (@1422115)</span></span></span>
<span><span class="co">##  BiocManager    1.30.25    <span style="color: #555555;">2024-08-28</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  BiocStyle    * 2.34.0     <span style="color: #555555;">2025-03-07</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">bioc_git2r (@0a04242)</span></span></span>
<span><span class="co">##  bitops         1.0-9      <span style="color: #555555;">2024-10-03</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  bookdown       0.42       <span style="color: #555555;">2025-01-07</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  bootstrap      <span style="color: #BB00BB; font-weight: bold;">2019.6    </span> <span style="color: #555555;">2019-06-17</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  bslib          0.9.0      <span style="color: #555555;">2025-01-30</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  cachem         1.1.0      <span style="color: #555555;">2024-05-16</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  caTools        1.18.3     <span style="color: #555555;">2024-09-04</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  cli            3.6.4      <span style="color: #555555;">2025-02-13</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  codetools      0.2-20     <span style="color: #555555;">2024-03-31</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  colorspace     2.1-1      <span style="color: #555555;">2024-07-26</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  data.table     1.17.0     <span style="color: #555555;">2025-02-22</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  desc           1.4.3      <span style="color: #555555;">2023-12-10</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  devtools       2.4.5      <span style="color: #555555;">2022-10-11</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  digest         0.6.37     <span style="color: #555555;">2024-08-19</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  dplyr          1.1.4      <span style="color: #555555;">2023-11-17</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  ellipsis       0.3.2      <span style="color: #555555;">2021-04-29</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  evaluate       1.0.3      <span style="color: #555555;">2025-01-10</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  farver         2.1.2      <span style="color: #555555;">2024-05-13</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  fastmap        1.2.0      <span style="color: #555555;">2024-05-15</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  foreach        1.5.2      <span style="color: #555555;">2022-02-02</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  formatR      * 1.14       <span style="color: #555555;">2023-01-17</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  fs             1.6.5      <span style="color: #555555;">2024-10-30</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  future         1.34.0     <span style="color: #555555;">2024-07-29</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  future.apply   1.11.3     <span style="color: #555555;">2024-10-27</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  generics       0.1.3      <span style="color: #555555;">2022-07-05</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  ggplot2      * 3.5.1      <span style="color: #555555;">2024-04-23</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  glmnet       * 4.1-8      <span style="color: #555555;">2023-08-22</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  globals        0.16.3     <span style="color: #555555;">2024-03-08</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  glue           1.8.0      <span style="color: #555555;">2024-09-30</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  gplots       * 3.2.0      <span style="color: #555555;">2024-10-05</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  gtable         0.3.6      <span style="color: #555555;">2024-10-25</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  gtools         3.9.5      <span style="color: #555555;">2023-11-20</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  htmltools      0.5.8.1    <span style="color: #555555;">2024-04-04</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  htmlwidgets    1.6.4      <span style="color: #555555;">2023-12-06</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  httpuv         1.6.15     <span style="color: #555555;">2024-03-26</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  iterators      1.0.14     <span style="color: #555555;">2022-02-05</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  jquerylib      0.1.4      <span style="color: #555555;">2021-04-26</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  jsonlite       1.9.1      <span style="color: #555555;">2025-03-03</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  KernSmooth     2.23-26    <span style="color: #555555;">2025-01-01</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  knitr        * 1.49       <span style="color: #555555;">2024-11-08</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  labeling       0.4.3      <span style="color: #555555;">2023-08-29</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  later          1.4.1      <span style="color: #555555;">2024-11-27</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  lattice        0.22-6     <span style="color: #555555;">2024-03-20</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  lava           1.8.1      <span style="color: #555555;">2025-01-12</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  lifecycle      1.0.4      <span style="color: #555555;">2023-11-07</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  listenv        0.9.1      <span style="color: #555555;">2024-01-29</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  magrittr       2.0.3      <span style="color: #555555;">2022-03-30</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  MASS           7.3-65     <span style="color: #555555;">2025-02-28</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  Matrix       * 1.7-2      <span style="color: #555555;">2025-01-23</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  memoise        2.0.1      <span style="color: #555555;">2021-11-26</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  mime           0.12       <span style="color: #555555;">2021-09-28</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  miniUI         0.1.1.1    <span style="color: #555555;">2018-05-18</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  multtest     * 2.62.0     <span style="color: #555555;">2025-03-07</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">bioc_git2r (@b2a4744)</span></span></span>
<span><span class="co">##  munsell        0.5.1      <span style="color: #555555;">2024-04-01</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  parallelly     1.42.0     <span style="color: #555555;">2025-01-30</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  pillar         1.10.1     <span style="color: #555555;">2025-01-07</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  pkgbuild       1.4.6      <span style="color: #555555;">2025-01-16</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  pkgconfig      2.0.3      <span style="color: #555555;">2019-09-22</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  pkgdown        2.1.1      <span style="color: #555555;">2024-09-17</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  pkgload        1.4.0      <span style="color: #555555;">2024-06-28</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  prodlim      * <span style="color: #BB00BB; font-weight: bold;">2024.06.25</span> <span style="color: #555555;">2024-06-24</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  profvis        0.4.0      <span style="color: #555555;">2024-09-20</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  promises       1.3.2      <span style="color: #555555;">2024-11-28</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  purrr          1.0.4      <span style="color: #555555;">2025-02-05</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  R6             2.6.1      <span style="color: #555555;">2025-02-15</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  ragg           1.3.3      <span style="color: #555555;">2024-09-11</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  rbibutils      2.3        <span style="color: #555555;">2024-10-04</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  Rcpp           1.0.14     <span style="color: #555555;">2025-01-12</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  Rdpack       * 2.6.2      <span style="color: #555555;">2024-11-15</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  remotes        2.5.0      <span style="color: #555555;">2024-03-17</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  rlang          1.1.5      <span style="color: #555555;">2025-01-17</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  rmarkdown      2.29       <span style="color: #555555;">2024-11-04</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  rmeta          3.0        <span style="color: #555555;">2018-03-20</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  ROCR         * 1.0-11     <span style="color: #555555;">2020-05-02</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  rstudioapi     0.17.1     <span style="color: #555555;">2024-10-22</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  sass           0.4.9      <span style="color: #555555;">2024-03-15</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  scales       * 1.3.0      <span style="color: #555555;">2023-11-28</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  scrime         1.3.5      <span style="color: #555555;">2018-12-01</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  sessioninfo    1.2.3      <span style="color: #555555;">2025-02-05</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  shape          1.4.6.1    <span style="color: #555555;">2024-02-23</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  shiny          1.10.0     <span style="color: #555555;">2024-12-14</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  siggenes     * 1.80.0     <span style="color: #555555;">2025-03-07</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">bioc_git2r (@18cb74e)</span></span></span>
<span><span class="co">##  SuppDists      1.1-9.8    <span style="color: #555555;">2024-09-03</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  survcomp     * 1.56.0     <span style="color: #555555;">2025-03-07</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">bioc_git2r (@1e00a05)</span></span></span>
<span><span class="co">##  survival     * 3.8-3      <span style="color: #555555;">2024-12-17</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  survivalROC    1.0.3.1    <span style="color: #555555;">2022-12-05</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  systemfonts    1.2.1      <span style="color: #555555;">2025-01-20</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  textshaping    0.3.7      <span style="color: #555555;">2023-10-09</span> <span style="color: #555555;">[4]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">##  tibble         3.2.1      <span style="color: #555555;">2023-03-20</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  tidyselect     1.2.1      <span style="color: #555555;">2024-03-11</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  urlchecker     1.0.1      <span style="color: #555555;">2021-11-30</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  usethis        3.1.0      <span style="color: #555555;">2024-11-26</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  vctrs          0.6.5      <span style="color: #555555;">2023-12-01</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  withr          3.0.2      <span style="color: #555555;">2024-10-28</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  xfun           0.51       <span style="color: #555555;">2025-02-19</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  xtable         1.8-4      <span style="color: #555555;">2019-04-21</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">##  yaml           2.3.10     <span style="color: #555555;">2024-07-26</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.4.3)</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #555555;"> [1] /tmp/Rtmp177AcV/temp_libpath79bc312bff30c</span></span></span>
<span><span class="co">## <span style="color: #555555;"> [2] /home/alberto/R/x86_64-pc-linux-gnu-library/4.4</span></span></span>
<span><span class="co">## <span style="color: #555555;"> [3] /usr/local/lib/R/site-library</span></span></span>
<span><span class="co">## <span style="color: #555555;"> [4] /usr/lib/R/site-library</span></span></span>
<span><span class="co">## <span style="color: #555555;"> [5] /usr/lib/R/library</span></span></span>
<span><span class="co">##  <span style="color: #BBBBBB; background-color: #BB0000;">*</span> ── Packages attached to the search path.</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB; font-weight: bold;">──────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-aguirre2013survexpress" class="csl-entry">
Aguirre-Gamboa, R, H Gomez-Rueda, E Martínez-Ledesma, A
Martínez-Torteya, R Chacolla-Huaringa, A Rodriguez-Barrientos, JG
Tamez-Peña, and V. Treviño. 2013. <span>“SurvExpress: An Online
Biomarker Validation Tool and Database for Cancer Gene Expression Data
Using Survival Analysis.”</span> <em>PLoS One</em> 8 (9): e74250. <a href="https://doi.org/10.1371/journal.pone.0074250" class="external-link">https://doi.org/10.1371/journal.pone.0074250</a>.
</div>
<div id="ref-barnett2008estrogen" class="csl-entry">
Barnett, DH, S Sheng, TH Charn, A Waheed, WS Sly, CY Lin, ET Liu, and BS
Katzenellenbogen. 2008. <span>“Estrogen Receptor Regulation of Carbonic
Anhydrase XII Through a Distal Enhancer in Breast Cancer.”</span>
<em>Cancer Research</em> 68 (9): 3505–15. <a href="https://doi.org/10.1158/0008-5472.CAN-07-6151" class="external-link">https://doi.org/10.1158/0008-5472.CAN-07-6151</a>.
</div>
<div id="ref-Bueno2023" class="csl-entry">
Bueno-Fortes, S, A Berral-Gonzalez, JM Sanchez-Santos, M Martin-Merino,
and J De Las Rivas. 2023. <span>“Identification of a Gene Expression
Signature Associated with Breast Cancer Survival and Risk That Improves
Clinical Genomic Platforms.”</span> <em>Bioinformatics Advances</em> 3
(1): vbad037. <a href="https://doi.org/10.1093/bioadv/vbad037" class="external-link">https://doi.org/10.1093/bioadv/vbad037</a>.
</div>
<div id="ref-glmnet" class="csl-entry">
Friedman, JH, T Hastie, and R Tibshirani. 2010. <span>“Regularization
Paths for Generalized Linear Models via Coordinate Descent.”</span>
<em>Journal of Statistical Software</em> 33 (1): 1–22. <a href="http://www.jstatsoft.org/v33/i01" class="external-link">http://www.jstatsoft.org/v33/i01</a>.
</div>
<div id="ref-garczyk2015agr3" class="csl-entry">
Garczyk, S, S VonStillfried, W Antonopoulos, A Hartmann, MG Schrauder,
PA Fasching, T Anzeneder, et al. 2015. <span>“AGR3 in Breast Cancer:
Prognostic Impact and Suitable Serum-Based Biomarker for Early Cancer
Detection.”</span> <em>PloS One</em> 10 (4): e0122106. <a href="https://doi.org/10.1371/journal.pone.0122106" class="external-link">https://doi.org/10.1371/journal.pone.0122106</a>.
</div>
<div id="ref-RMA" class="csl-entry">
Gautier, L, L Cope, BM Bolstad, and RA Irizarry. 2004.
<span>“Affy—Analysis of Affymetrix GeneChip Data at the Probe
Level.”</span> <em>Bioinformatics</em> 20 (3): 307–15. <a href="https://doi.org/10.1093/bioinformatics/btg405" class="external-link">https://doi.org/10.1093/bioinformatics/btg405</a>.
</div>
<div id="ref-kassambara2017package" class="csl-entry">
Kassambara, A, M Kosinski, P Biecek, and S Fabian. 2024.
<span>“Survminer: Drawing Survival Curves Using Ggplot2 (r Package
Version 0.5.0).”</span> <em>CRAN-R</em>. <a href="https://doi.org/10.32614/CRAN.package.survminer" class="external-link">https://doi.org/10.32614/CRAN.package.survminer</a>.
</div>
<div id="ref-combat" class="csl-entry">
Kupfer, P, R Guthke, D Pohlers, R Huber, D Koczan, and RW Kinne. 2012.
<span>“<span class="nocase"><span>B</span>atch correction of microarray
data substantially improves the identification of genes differentially
expressed in rheumatoid arthritis and osteoarthritis</span>.”</span>
<em>BMC Medical Genomics</em> 5: 23. <a href="https://doi.org/10.1186/1755-8794-5-23" class="external-link">https://doi.org/10.1186/1755-8794-5-23</a>.
</div>
<div id="ref-li2013finding" class="csl-entry">
Li, J, and R Tibshirani. 2013. <span>“Finding Consistent Patterns: A
Nonparametric Approach for Identifying Differential Expression in
RNA-Seq Data.”</span> <em>Statistical Methods in Medical Research</em>
22 (5): 519–36. <a href="https://doi.org/10.1177/0962280211428386" class="external-link">https://doi.org/10.1177/0962280211428386</a>.
</div>
<div id="ref-fRMA" class="csl-entry">
McCall, MN, BM Bolstad, and RA Irizarry. 2010. <span>“Frozen Robust
Multiarray Analysis (fRMA).”</span> <em>Biostatistics</em> 11 (2):
242–53. <a href="https://doi.org/10.1093/biostatistics/kxp059" class="external-link">https://doi.org/10.1093/biostatistics/kxp059</a>.
</div>
<div id="ref-ritchie2015limma" class="csl-entry">
Ritchie, ME, B Phipson, D Wu, Y Hu, CW Law, W Shi, and GK Smyth. 2015.
<span>“Limma Powers Differential Expression Analyses for RNA-Sequencing
and Microarray Studies.”</span> <em>Nucleic Acids Research</em> 43 (7):
e47–47. <a href="https://doi.org/10.1093/nar/gkv007" class="external-link">https://doi.org/10.1093/nar/gkv007</a>.
</div>
<div id="ref-scarfo2016identification" class="csl-entry">
Scarfo, I, E Pellegrino, E Mereu, I Kwee, L Agnelli, E Bergaggio, G
Garaffo, et al. 2016. <span>“Identification of a New Subclass of
ALK-Negative ALCL Expressing Aberrant Levels of ERBB4
Transcripts.”</span> <em>Blood</em> 127 (2): 221–32. <a href="https://doi.org/10.1182/blood-2014-12-614503" class="external-link">https://doi.org/10.1182/blood-2014-12-614503</a>.
</div>
<div id="ref-simon2011regularization" class="csl-entry">
Simon, N, J Friedman, T Hastie, and R Tibshirani. 2011.
<span>“Regularization Paths for Cox’s Proportional Hazards Model via
Coordinate Descent.”</span> <em>Journal of Statistical Software</em> 39
(5): 1. <a href="https://doi.org/10.18637/jss.v039.i05" class="external-link">https://doi.org/10.18637/jss.v039.i05</a>.
</div>
<div id="ref-survival-package" class="csl-entry">
Therneau, TM. 2021. <span>“A Package for Survival Analysis in r.”</span>
<em>CRAN-R</em>. <a href="https://CRAN.R-project.org/package=survival" class="external-link">https://CRAN.R-project.org/package=survival</a>.
</div>
<div id="ref-tibshirani2009univariate" class="csl-entry">
Tibshirani, R. 2009. <span>“Univariate Shrinkage in the Cox Model for
High Dimensional Data.”</span> <em>Statistical Applications in Genetics
and Molecular Biology</em> 8 (1). <a href="https://doi.org/10.2202/1544-6115.1438" class="external-link">https://doi.org/10.2202/1544-6115.1438</a>.
</div>
<div id="ref-tibshirani1996regression" class="csl-entry">
Tibshirani, Robert. 1996. <span>“Regression Shrinkage and Selection via
the Lasso.”</span> <em>Journal of the Royal Statistical Society: Series
B (Methodological)</em> 58 (1): 267–88. <a href="http://www.jstor.org/stable/2346178" class="external-link">http://www.jstor.org/stable/2346178</a>.
</div>
<div id="ref-Tusher24042001" class="csl-entry">
Tusher, VG, R Tibshirani, and G Chu. 2001. <span>“Significance Analysis
of Microarrays Applied to the Ionizing Radiation Response.”</span>
<em>Proceedings of the National Academy of Sciences</em> 98 (9):
5116–21. <a href="https://doi.org/10.1073/pnas.091062498" class="external-link">https://doi.org/10.1073/pnas.091062498</a>.
</div>
<div id="ref-yu2015sushi" class="csl-entry">
Yu, Z, E Jiang, X Wang, Y Shi, AJ Shangguan, L Zhang, and J Li. 2015.
<span>“Sushi Domain-Containing Protein 3: A Potential Target for Breast
Cancer.”</span> <em>Cell Biochemistry and Biophysics</em> 72 (2):
321–24. <a href="https://doi.org/10.1007/s12013-014-0480-9" class="external-link">https://doi.org/10.1007/s12013-014-0480-9</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alberto Berral-Gonzalez, Santiago Bueno-Fortes, Natalia Alonso-Moreda, Jose Manuel Sanchez-Santos, Manuel Martin-Merino Acera, Javier De Las Rivas.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
