<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>geneSurv — geneSurv • asuri</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="geneSurv — geneSurv"><meta name="description" content="This function analyzes the ability of a gene to mark survival based on a robust version of the KM curves. The robust K-M estimator is obtained by a bootstrap strategy."><meta property="og:description" content="This function analyzes the ability of a gene to mark survival based on a robust version of the KM curves. The robust K-M estimator is obtained by a bootstrap strategy."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">asuri</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/asuri.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jdelasrivas-lab/asuri/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>geneSurv</h1>

      <div class="d-none name"><code>geneSurv.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function analyzes the ability of a gene to mark survival based on a robust version of the KM curves. The robust K-M estimator is obtained by a bootstrap strategy.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">geneSurv</span><span class="op">(</span><span class="va">genExpr</span>, <span class="va">time</span>, <span class="va">status</span>, <span class="va">geneName</span>, boxplot <span class="op">=</span> <span class="cn">T</span>, iter <span class="op">=</span> <span class="fl">100</span>, type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"exprs"</span>, <span class="st">"risk"</span><span class="op">)</span>, cut_time <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <p></p>
<dl><dt id="arg-genexpr">genExpr<a class="anchor" aria-label="anchor" href="#arg-genexpr"></a></dt>
<dd><p>Vector with the normalized gene expression for each sample. names(genExpr) should contains the sample names.</p></dd>

  <dt id="arg-time">time<a class="anchor" aria-label="anchor" href="#arg-time"></a></dt>
<dd><p>Numeric vector containing the survival time for each sample in years, including the sample name in names(time).</p></dd>

  <dt id="arg-status">status<a class="anchor" aria-label="anchor" href="#arg-status"></a></dt>
<dd><p>Numeric vector with the status (censored 0 and not censored 1) for each sample. names(status) shoud include the sample names.</p></dd>

  <dt id="arg-genename">geneName<a class="anchor" aria-label="anchor" href="#arg-genename"></a></dt>
<dd><p>A character string with the name of the gene being analyzed.</p></dd>

   <dt id="arg-boxplot">boxplot<a class="anchor" aria-label="anchor" href="#arg-boxplot"></a></dt>
<dd><p>A logical value indicating whether to generate a boxplot of gene expression by survival group (default = TRUE).</p></dd>

   <dt id="arg-iter">iter<a class="anchor" aria-label="anchor" href="#arg-iter"></a></dt>
<dd><p>The number of iterations (bootstrap resampling) for calculating optimal group cutoffs (default = 100).</p></dd>

   <dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>Defines if the KM curve groups are computed using risk ("risk") or gene expression (default "exprs").</p></dd>

   <dt id="arg-cut-time">cut_time<a class="anchor" aria-label="anchor" href="#arg-cut-time"></a></dt>
<dd><p>A numeric value specifying the cutoff time (in years) for survival analysis. All events beyond this time are treated as censored (default = 10 years).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function improves the stability and robustness of the K-M estimator using a bootstrap strategy. Patients are resampled with replacement giving rise to B replicates. The K-M estimator is obtained based on the replicates as well as the confidence intervals. The patients are stratified in two risk groups by an expression threshold that optimizes the log-rank statistics, that is the separability between the Kaplan-Meier curves for each group. This function implements a novel method to find the optimal threshold avoiding the problems of instability and unbalanced classes that suffer other implementations.
Besides, a membership probability for each risk group is estimated from the classification of each sample in the replicates. This membership probability allow us to reclassify patients around the gene expression threshold in a more robust way.
The function provides a robust estimation of the log-rank p-value and the Hazard ratio that allow us to evaluate the ability of a given gene to mark survival.</p>
    </div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Depending on the type run, the output changes.</p><dl><dt>For "type = exprs":</dt>
<dd><p>A Kaplan-Meier plot based on expression groups, a differential expression boxplot and a plot with the membership probability for each risk group. Additionally, an object with the following components:</p><dl><dt>geneName</dt>
<dd><p>A character string with the name of the gene being analyzed.</p></dd>

      <dt>patientExpr</dt>
<dd><p>The expression level of each patient for the given gene.</p></dd>

      <dt>patientClass</dt>
<dd><p>Vector of group classification according to the gene expression level: 2 = high expression and 1 = low expression level.</p></dd>

      <dt>patientClassProbality</dt>
<dd><p>Vector of membership probabilities for the classification.</p></dd>

      <dt>wilcox.pvalue</dt>
<dd><p>The p-value from the Wilcoxon test comparing the two expression groups.</p></dd>

      <dt>plot_values</dt>
<dd><p>A list containing Kaplan-Meier fit results, log-rank p-value, and hazard ratio.</p></dd>


</dl><p></p></dd>

    <dt>For "type = risk":</dt>
<dd><p>A Kaplan-Meier plot based on risk groups. Additionally, an object with the following components:</p><dl><dt>geneName</dt>
<dd><p>A character string with the name of the gene being analyzed.</p></dd>

        <dt>patientExpr</dt>
<dd><p>The expression level of each patient for the given gene.</p></dd>

        <dt>risk_score_predicted</dt>
<dd><p>A numeric vector of predicted relative risk scores for each patient.</p></dd>

        <dt>plot_values</dt>
<dd><p>A list containing Kaplan-Meier fit results, log-rank p-value, and hazard ratio.</p></dd>


</dl><p></p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Martinez-Romero, J., Bueno-Fortes, S., Martín-Merino, M. <em>et al.</em> (2018) <b>Survival marker genes of colorectal cancer derived from consistent transcriptomic profiling</b>. <em>BMC Genomics</em> 19 (Suppl 8), 857 . <a href="https://doi.org/10.1186/s12864-018-5193-9" class="external-link">doi:10.1186/s12864-018-5193-9</a></p>
<p>Santiago Bueno-Fortes, Alberto Berral-Gonzalez, José Manuel Sánchez-Santos, Manuel Martin-Merino, Javier De Las Rivas (2023) <b>Identification of a gene expression signature associated with breast cancer survival and risk that improves clinical genomic platforms</b>, <em>Bioinformatics Advances</em>, Volume 3, Issue 1, vbad037, <a href="https://doi.org/10.1093/bioadv/vbad037" class="external-link">doi:10.1093/bioadv/vbad037</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">matrixData</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>data set ‘matrixData’ not found</span>
<span class="r-in"><span>  </span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">time</span> <span class="op">&lt;-</span> <span class="va">mPheno</span><span class="op">$</span><span class="va">time</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> object 'mPheno' not found</span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">time</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mPheno</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in h(simpleError(msg, call)):</span> error in evaluating the argument 'x' in selecting a method for function 'rownames': object 'mPheno' not found</span>
<span class="r-in"><span>  <span class="va">status</span> <span class="op">&lt;-</span> <span class="va">mPheno</span><span class="op">$</span><span class="va">status</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> object 'mPheno' not found</span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">status</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mPheno</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in h(simpleError(msg, call)):</span> error in evaluating the argument 'x' in selecting a method for function 'rownames': object 'mPheno' not found</span>
<span class="r-in"><span>  <span class="co"># The TIME value must be transformed to YEARS</span></span></span>
<span class="r-in"><span>  <span class="co"># The gene expression vector must be provided with the NAMES of each sample, that should match with the NAMES of the time and status vectors </span></span></span>
<span class="r-in"><span>  <span class="va">outputKM</span> <span class="op">&lt;-</span> <span class="fu">geneSurv</span><span class="op">(</span><span class="va">mExprs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="st">"ESR1"</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mExprs</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span>, <span class="va">time</span>, <span class="va">status</span>, <span class="st">"ESR1"</span>, </span></span>
<span class="r-in"><span>                       type <span class="op">=</span> <span class="st">"exprs"</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> object 'mExprs' not found</span>
<span class="r-in"><span>  </span></span>
<span class="r-in"><span>  <span class="co"># If we consider p53 gene instead (ENSG00000141510)</span></span></span>
<span class="r-in"><span>  <span class="va">outputKM.TP53</span> <span class="op">&lt;-</span> <span class="fu">geneSurv</span><span class="op">(</span><span class="va">mExprs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="st">"TP53"</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mExprs</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span>, <span class="va">time</span>, <span class="va">status</span>, <span class="st">"TP53"</span>, </span></span>
<span class="r-in"><span>                            type <span class="op">=</span> <span class="st">"exprs"</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> object 'mExprs' not found</span>
<span class="r-in"><span>                            </span></span>
<span class="r-in"><span>  <span class="co"># Generate the plots again</span></span></span>
<span class="r-in"><span>  <span class="co">## Plots for c(type = exprs)</span></span></span>
<span class="r-in"><span>  <span class="fu">asuri</span><span class="fu">:::</span><span class="fu">plotBoxplot</span><span class="op">(</span><span class="va">outputKM</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> object 'outputKM' not found</span>
<span class="r-in"><span>  <span class="fu">asuri</span><span class="fu">:::</span><span class="fu">plotProbClass</span><span class="op">(</span><span class="va">outputKM</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> object 'outputKM' not found</span>
<span class="r-in"><span>  <span class="fu">asuri</span><span class="fu">:::</span><span class="fu">plotKM</span><span class="op">(</span><span class="va">outputKM</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> object 'outputKM' not found</span>
<span class="r-in"><span>  <span class="co">## Plots for c(type = risk)</span></span></span>
<span class="r-in"><span>  <span class="fu">asuri</span><span class="fu">:::</span><span class="fu">plotKM</span><span class="op">(</span><span class="va">outputKM</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> object 'outputKM' not found</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alberto Berral-Gonzalez, Santiago Bueno-Fortes, Natalia Alonso-Moreda, Jose Manuel Sanchez-Santos, Manuel Martin-Merino Acera, Javier De Las Rivas.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

