<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>prefilterSAM — prefilterSAM • asuri</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="prefilterSAM — prefilterSAM"><meta name="description" content="R function to pre-filter noisy genes, keeping the most relevant ones for microarrays or RNAseq data. These genes can be used later for further analysis."><meta property="og:description" content="R function to pre-filter noisy genes, keeping the most relevant ones for microarrays or RNAseq data. These genes can be used later for further analysis."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">asuri</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/asuri.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jdelasrivas-lab/asuri/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>prefilterSAM</h1>

      <div class="d-none name"><code>prefilterSAM.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>R function to pre-filter noisy genes, keeping the most relevant ones for microarrays or RNAseq data. These genes can be used later for further analysis.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">prefilterSAM</span><span class="op">(</span><span class="va">mExpr</span>, <span class="va">groups_vector</span>, FDRfilter <span class="op">=</span> <span class="fl">0.05</span>, percentageFilter <span class="op">=</span> <span class="fl">80</span>, iter <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <p></p>
<dl><dt id="arg-mexpr">mExpr<a class="anchor" aria-label="anchor" href="#arg-mexpr"></a></dt>
<dd><p>A matrix where rows represent genes and columns represent samples. Each entry is the expression value of a gene in a sample.</p></dd>

  <dt id="arg-groups-vector">groups_vector<a class="anchor" aria-label="anchor" href="#arg-groups-vector"></a></dt>
<dd><p>A binary vector indicating the group assignment for each sample.</p></dd>

  <dt id="arg-fdrfilter">FDRfilter<a class="anchor" aria-label="anchor" href="#arg-fdrfilter"></a></dt>
<dd><p>A numeric value indicating the FDR threshold for selecting significant genes. Default is 0.05.</p></dd>

  <dt id="arg-iter">iter<a class="anchor" aria-label="anchor" href="#arg-iter"></a></dt>
<dd><p>The number of iterations for bootstrapping. Default is 100.</p></dd>

  <dt id="arg-percentagefilter">percentageFilter<a class="anchor" aria-label="anchor" href="#arg-percentagefilter"></a></dt>
<dd><p>A numeric value indicating the percentage of iterations a gene must appear in to be considered significant. Default is 80.</p></dd>

  </dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function implements SAM (Schwender H., 2022) robust diferential expression analysis based on bootstrap . It helps to remove noisy genes reducing the computational complexity of further analysis.</p>
    </div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>An ordered vector with the names of differentially expressed genes between the categories of the grouping vector.</p>
<dl><dt>result</dt>
<dd><p>A list of DE genes ordered by SAM d.value and filtered by percentageFilter.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Schwender H (2022). <b>_siggenes: Multiple Testing using SAM and Efron's Empirical Bayes Approaches_.</b> R package version 1.72.0.</p>
<p>Martinez-Romero, J., Bueno-Fortes, S., Martín-Merino, M. <em>et al.</em> (2018) <b>Survival marker genes of colorectal cancer derived from consistent transcriptomic profiling</b>. <em>BMC Genomics</em> 19 (Suppl 8), 857 . <a href="https://doi.org/10.1186/s12864-018-5193-9" class="external-link">doi:10.1186/s12864-018-5193-9</a></p>
<p>Santiago Bueno-Fortes, Alberto Berral-Gonzalez, José Manuel Sánchez-Santos, Manuel Martin-Merino, Javier De Las Rivas (2023) <b>Identification of a gene expression signature associated with breast cancer survival and risk that improves clinical genomic platforms</b>, <em>Bioinformatics Advances</em>, Volume 3, Issue 1, vbad037, <a href="https://doi.org/10.1093/bioadv/vbad037" class="external-link">doi:10.1093/bioadv/vbad037</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">matrixData</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>data set ‘matrixData’ not found</span>
<span class="r-in"><span>  <span class="co"># Bootstrapped differential expression based on SAM. Parameters: FDR = 0.05, </span></span></span>
<span class="r-in"><span>  <span class="co"># iter = 100, percentage filter = 80 % </span></span></span>
<span class="r-in"><span>  <span class="co"># CAUTION: if the data have a high number of genes this function will take several minutes to compute.</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">DE_list_genes</span> <span class="op">&lt;-</span> <span class="fu">prefilterSAM</span><span class="op">(</span><span class="va">mExprs</span><span class="op">[</span>,<span class="op">]</span>, <span class="va">mPheno</span><span class="op">$</span><span class="va">ER.IHC</span><span class="op">)</span> </span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> object 'mExprs' not found</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alberto Berral-Gonzalez, Santiago Bueno-Fortes, Natalia Alonso-Moreda, Jose Manuel Sanchez-Santos, Manuel Martin-Merino Acera, Javier De Las Rivas.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

